Userland/DevTools/Profiler/SourceModel.cpp

```diff
@@ -0,0 +1,221 @@
+/*
+ * Copyright (c) 2021, Stephan Unverwerth <s.unverwerth@serenityos.org>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include "SourceModel.h"
+#include "Profile.h"
+#include <LibDebug/DebugInfo.h>
+#include <LibGUI/Painter.h>
+#include <LibSymbolication/Symbolication.h>
+#include <stdio.h>
+
+namespace Profiler {
+
+static constexpr StringView c_source_root_path = "/usr/src/serenity/"sv;
+
+class SourceFile final {
+public:
+    struct Line {
+        String content;
+        size_t num_samples { 0 };
+    };
+
+public:
+    SourceFile(StringView filename)
+    {
+        String source_file_name = filename.replace("../../", c_source_root_path);
+
+        auto maybe_file = Core::File::open(source_file_name, Core::OpenMode::ReadOnly);
+        if (maybe_file.is_error()) {
+            dbgln("Could not map source file \"{}\". Tried {}. {} (errno={})", filename, source_file_name, maybe_file.error().string_literal(), maybe_file.error().code());
+            return;
+        }
+
+        auto file = maybe_file.value();
+
+        while (!file->eof()) {
+            m_lines.append({ file->read_line(1024), 0 });
+        }
```

Comment:
redundant curly braces

```suggestion
        while (!file->eof())
            m_lines.append({ file->read_line(1024), 0 });
```